server/balancer.go

```diff
@@ -370,3 +372,272 @@ func (r *replicaChecker) checkBestReplacement(region *RegionInfo) Operator {
+		}
+		l.scoreStatus[sid] = status
+	}
+}
+func (l *balanceHotRegionScheduler) CalculateScore(cluster *clusterInfo) {
+	l.clearScore()
+	items := cluster.writeStatistics.elems()
+	for _, item := range items {
+		r, ok := item.value.(RegionStat)
+		if !ok {
+			continue
+		}
+		if r.HotDegree < hotRegionLowThreshold {
+			continue
+		}
+
+		id := r.RegionID
+		regionInfo := cluster.getRegion(id)
+		storeID := regionInfo.Leader.GetStoreId()
+		status, ok := l.scoreStatus[storeID]
+		if !ok {
+			status = StoreHotRegions{
+				RegionsStat: make(ListRegionsStat, 0, 100),
+			}
+			l.scoreStatus[storeID] = status
+		}
+		status.StoreTotalWrittenBytes += r.WrittenBytes
+		status.RegionsStat = append(status.RegionsStat, RegionStat{id, r.WrittenBytes, r.HotDegree, r.LastUpdateTime, storeID, r.antiCount, r.version})
+		//l.scoreStatus[storeID] = status
+	}
+
+	for _, rs := range l.scoreStatus {
+		sort.Sort(sort.Reverse(rs.RegionsStat))
+	}
+}
+
+func (l *balanceHotRegionScheduler) SelectSourceRegion(cluster *clusterInfo) *RegionInfo {
+	var (
+		maxWritten               uint64
+		sourceStore              uint64
+		maxHotStoreRegionsNumber int
+	)
+	// choose a hot store as transfer source
+	// the numbers of the hot regions in that store is considered priority than StoreTotalWrittenBytes
+	for sid, s := range l.scoreStatus {
+		if s.RegionsStat.Len() < 2 {
+			continue
+		}
+
+		if maxHotStoreRegionsNumber < s.RegionsStat.Len() {
+			maxHotStoreRegionsNumber = s.RegionsStat.Len()
+			maxWritten = s.StoreTotalWrittenBytes
+			sourceStore = sid
+			continue
+		}
+
+		if maxHotStoreRegionsNumber == s.RegionsStat.Len() && maxWritten < s.StoreTotalWrittenBytes {
+			maxWritten = s.StoreTotalWrittenBytes
+			sourceStore = sid
+		}
+	}
+
+	if sourceStore == 0 {
+		return nil
+	}
+
+	length := l.scoreStatus[sourceStore].RegionsStat.Len()
+	// radmon pick a region from 1 .. length-1
+	// TODO: consider hot degree when pick
+	rr := l.r.Int31n(int32(length-1)) + 1
+	s := l.scoreStatus[sourceStore].RegionsStat[rr]
+	if s.antiCount < hotRegionAntiCount {
+		return nil
+	}
+	region := cluster.getRegion(s.RegionID)
+	if len(region.DownPeers) != 0 || len(region.PendingPeers) != 0 {
+		return nil
+	}
+	// use written bytes per second
+	region.WrittenBytes = s.WrittenBytes
+	l.adjustBalanceLimit(sourceStore)
+	return region
+}
+
+func (l *balanceHotRegionScheduler) adjustBalanceLimit(storeID uint64) {
+	s := l.scoreStatus[storeID]
+	var regionTotalCount float64
+	for _, m := range l.scoreStatus {
+		regionTotalCount += float64(m.RegionsStat.Len())
+	}
+
+	avgRegionCount := regionTotalCount / float64(len(l.scoreStatus))
+	limit := uint64((float64(s.RegionsStat.Len()) - avgRegionCount) / 2)
+	l.limit = maxUint64(1, limit)
+}
+
+func (l *balanceHotRegionScheduler) GetStatus() map[uint64]StoreHotRegions {
+	return l.scoreStatus
+}
+
+func (l *balanceHotRegionScheduler) SelectTransferLeader(cluster *clusterInfo) (*RegionInfo, *metapb.Peer) {
```

Comment:
This function do two different things(select source region and select target leader), the function name make me confused.